File: C++/TwoSum.cc
// Given an array of integers, find two numbers such that they add up to a specific target number.
// The function twoSum should return indices of the two numbers such that they add up to the target, 
// where index1 must be less than index2. Please note that your returned answers(both index1 and index2) 
// are not zero-based.
// Assume that each input would have exactly one solution

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // key is the number in nums, value is the corresponding index
        unordered_map<int, int> hashtable;
        vector<int> result;
        int size = nums.size();
        for (int i = 0; i < size; ++i) {
            auto it = hashtable.find(target - nums[i]);
            if(it != hashtable.end()) {
                result = { it->second + 1, i + 1 };
                break;
            } else {
                hashtable.emplace(nums[i], i);
            }
        }
        return result;
    }
};

File: Go/two_sum.go
// Given an array of integers, return indices of the two numbers such that they add up to a specific target.

// You may assume that each input would have exactly one solution.

// Example:
// Given nums = [2, 7, 11, 15], target = 9,

// Because nums[0] + nums[1] = 2 + 7 = 9,
// return [0, 1].

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)
	for i, v := range nums {
		j, ok := m[target - v]
		if ok {
			return []int{j, i}
		} else {
			m[v] = i
		}
	}
	return []int{-1, -1}
}
